content/blog_posts/articles/hmr-in-adonisjs.md

```diff
@@ -0,0 +1,106 @@
+
+However, with TSX: we're working with JavaScript modules. And, when a JavaScript module is imported, it gets cached. No matter how many times you modify the file, the imported module will always remain the same until the server is restarted.
+
+As a result, even a simple modification of padding/margin in our TSX view would cause a complete server restart, which can be long and tedious when developing your UI, for example.
+
+That's where HMR comes in: now, a modification of our TSX view just causes a page refresh, without server restart, and it's almost instantaneous. We now have an even better DX with TSX. **Really, it's a game changer.**
+
+### Large applications
+
+For applications that use a lot of packages (and thus `providers`), server restarts can be slows. The more providers you add, the more code there is to import and initialize at application startup. And the boot time can quickly reach several seconds: again, for the most impatient among us, this can be a bit frustrating development experience.
+
+### ESM slower than CommonJS
+
+To accentuate the problem mentioned above, there's also the fact that ESM loads significantly slower than CommonJS. Bun talks about this in this [blog post](https://bun.sh/blog/commonjs-is-not-going-away#the-case-for-commonjs):
+
+![esm-vs-cjs](./cjs-vs-esm.png)
+
+The benchmark is done with Bun, but with Node.JS the numbers are similar. The Node.js team is actively working on improving ESM performance, but for now, the fact is that ESM loads slower than CommonJS.
+
+As a side note, this is also why most Adonis.JS packages are bundled with `tsup`: it means fewer files to load, and thus a really faster boot time. During the development of AdonisJS 6 last summer, we managed to cut the application boot time in half just by bundling all of our packages.
+
+### Vite Experimental
+
+Finally, as we explained in [this article](https://adonisjs.com/blog/future-plans-for-adonisjs-6#current-implementation), our approach with Vite has changed: to have SSR support, we must launch Vite directly in the AdonisJS process rather than running it as a subprocess. 
+
+But this also impacted the application boot time, obviously, because we must relaunch Vite each time the server is restarted.
+
+## Conclusion
+
+These are the main reasons we added HMR to AdonisJS. We've tested this feature internally for several weeks, in real conditions, and we are super happy with the outcomes.
+
+We hardly ever need to reload the whole server now: only for things like adding a route or modifying a config file. But that's a minor part of the development process. Most of the time, it's all about HMR, and it really saves time and increases comfort.
```

Comment:
```suggestion

```

Reply:
why?

Reply:
Because this sentence is not needed in the article and it does not follow the tone of the article.